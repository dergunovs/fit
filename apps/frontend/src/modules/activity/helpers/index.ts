import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
