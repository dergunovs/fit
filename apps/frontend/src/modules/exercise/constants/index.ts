export const URL_EXERCISE = '/exercises';
export const URL_EXERCISE_CREATE = `${URL_EXERCISE}/create`;
export const URL_EXERCISE_EDIT = `${URL_EXERCISE}/edit`;

export const EXERCISE_REPEATS_OPTIONS = [8, 10, 12, 15, 20];
export const EXERCISE_REPEATS_DEFAULT = 12;

export const EXERCISE_STATISTICS_HEADERS = [{ title: 'Упражнение' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const MUSCLE_GROUP_STATISTICS_HEADERS = [{ title: 'Мышцы' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const EXERCISE_LIST_HEADERS = [{ title: 'Название' }];

export const EXERCISE_DELETED_TITLE = 'Упражнение удалено';
