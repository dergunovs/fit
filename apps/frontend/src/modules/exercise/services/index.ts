import { ComputedRef, Ref } from 'vue';
import {
  API_EXERCISE,
  API_EXERCISE_ALL,
  API_EXERCISE_CUSTOM,
  TGetExercisesDTO,
  TGetExercisesCustomDTO,
  TGetExercisesQueryDTO,
  TGetExerciseDTO,
  TPostExerciseDTO,
  TPostExerciseDataDTO,
  TUpdateExerciseDTO,
  TUpdateExerciseDataDTO,
  TDeleteExerciseDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const exerciseService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_EXERCISE, page],
      queryFn: async () => {
        const params: TGetExercisesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE, { params });

        return data;
      },
    });
  },

  getAll: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_ALL],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE_ALL);

        return data.data;
      },
    });
  },

  getCustom: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_CUSTOM],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesCustomDTO>(API_EXERCISE_CUSTOM);

        return data.data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_EXERCISE, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetExerciseDTO>(`${API_EXERCISE}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TPostExerciseDataDTO) => {
        const { data } = await api.post<TPostExerciseDTO>(API_EXERCISE, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TUpdateExerciseDataDTO) => {
        const { data } = await api.patch<TUpdateExerciseDTO>(`${API_EXERCISE}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteExerciseDTO>(`${API_EXERCISE}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
