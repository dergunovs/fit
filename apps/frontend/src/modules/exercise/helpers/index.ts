import {
  IExercise,
  IExerciseChoosen,
  IExerciseDone,
  IExerciseStatistics,
  IUser,
  IUserEquipment,
} from 'fitness-tracker-contracts';
import { createTempId } from 'mhz-helpers';

import { ITimelineStep } from '@/activity/interface';
import { getWeightsForUserEquipment } from '@/equipment/helpers';

export function getUserEquipmentParams(exercise: IExercise, user?: IUser) {
  const isExerciseHasEquipment = !!exercise.equipment;
  const isExerciseHasEquipmentForWeight = !!exercise.equipmentForWeight?.length;
  const isWeightsRequired = !!exercise.isWeightsRequired;

  const isUserHasEquipment = !!user?.equipments?.some(
    (equipment) => equipment.equipment?._id?.toString() === exercise.equipment?._id?.toString()
  );

  const isUserHasEquipmentForWeight = !!user?.equipments?.some((equipment) =>
    exercise.equipmentForWeight?.some(
      (equipmentForWeight) => equipmentForWeight._id?.toString() === equipment.equipment?._id?.toString()
    )
  );

  return {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  };
}

export function isUserEquipmentMatches(exercise: IExercise, user?: IUser) {
  let result = false;

  const {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  } = getUserEquipmentParams(exercise, user);

  if (!isExerciseHasEquipment && !isExerciseHasEquipmentForWeight) {
    result = true;
  } else if (isUserHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (!isExerciseHasEquipment && isExerciseHasEquipmentForWeight && isUserHasEquipmentForWeight) {
    result = true;
  } else if (!isExerciseHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (
    isExerciseHasEquipment &&
    isExerciseHasEquipmentForWeight &&
    isUserHasEquipment &&
    isUserHasEquipmentForWeight
  ) {
    result = true;
  }

  return result;
}

export function getAverageDuration(exercise: IExerciseStatistics, type: 'set' | 'repeat') {
  return type === 'set'
    ? `${((exercise.averageDuration * exercise.repeats) / exercise.sets || 0).toFixed(1)}`
    : `${exercise.averageDuration.toFixed(1)}`;
}

export function getAverageRepeatsInSet(repeats?: number, sets?: number) {
  return repeats && sets ? Math.round(repeats / sets) : 0;
}

export function isPrevExerciseSame(exercises: IExerciseDone[], index: number, id?: string) {
  return id && exercises[index - 1] ? id === exercises[index - 1].exercise?._id : false;
}

export function isSetCreatable(choosenExercises: IExerciseChoosen[], index: number, id?: string) {
  return !isPrevExerciseSame(choosenExercises, index, id) && index > 0 && index + 1 === choosenExercises.length;
}

export function generateTimeline(exercises: IExerciseDone[], start: Date | string | null, ratio: number) {
  const allSteps: ITimelineStep[] = [];

  exercises.forEach((exercise, index) => {
    if (!exercise.dateUpdated || !exercise.duration || !start) return;

    allSteps.push({
      left:
        index === 0
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - exercise.duration * 1000 - new Date(start).getTime()) / ratio,
      right:
        index === exercises.length - 1 && exercises.length !== 1
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - new Date(start).getTime()) / ratio,
      type: 'exercise',
    });
  });

  allSteps.forEach((step, index) => {
    if (!allSteps[index - 1]?.right) return;

    allSteps.push({ left: allSteps[index - 1].right, right: step.left, type: 'rest' });
  });

  return allSteps.sort((a, b) => a.left - b.left);
}

export function filterExercisesByTitleAndMuscle(exercises: IExercise[], title: string, muscle?: string, user?: IUser) {
  return exercises.filter((exercise) => {
    const isEquipmentMatches = isUserEquipmentMatches(exercise, user);

    const titleFilter =
      (exercise.title.toLowerCase().includes(title.toLocaleLowerCase()) ||
        exercise.title_en?.toLowerCase().includes(title.toLocaleLowerCase())) &&
      isEquipmentMatches;
    const muscleFilter = exercise.muscles?.some((group) => group._id === muscle) && isEquipmentMatches;

    return muscle ? muscleFilter && titleFilter : titleFilter;
  });
}

export function getAvailableExerciseWeights(exercise: IExercise, user: IUser) {
  const equipments = user.equipments?.filter((equipment) =>
    exercise.equipmentForWeight?.some((eq) => eq._id === equipment.equipment?._id)
  );

  if (!equipments?.length) return [];

  const weights = getWeightsForUserEquipment(equipments);

  return exercise.isWeightsRequired ? weights : [0, ...weights];
}

export function getExercisesToChooseDefaultWeight(exercises: IExercise[], userEquipments?: IUserEquipment[]) {
  const equipmentsForWeight = userEquipments?.filter((eq) => eq.weights?.length);
  const exercisesWithWeight = exercises.filter((ex) => ex.isWeights);

  const tableData = exercisesWithWeight.map((ex) => {
    const exerciseEquipment = ex.equipmentForWeight
      ?.filter((eq) => equipmentsForWeight?.some((equipment) => equipment.equipment?._id === eq._id))
      .map((exToMap) => exToMap._id);

    const availableEquipment = equipmentsForWeight?.filter((eq) =>
      exerciseEquipment?.some((equipmentId) => equipmentId === eq.equipment?._id)
    );

    const options = getWeightsForUserEquipment(availableEquipment);

    return {
      _id: ex._id,
      title: ex.title,
      title_en: ex.title_en,
      options: ex.isWeightsRequired ? options : [0, ...options],
    };
  });

  return tableData.filter((ex) => ex.options.length > 1);
}

export function getDefaultExerciseWeight(exercise: IExercise, user: IUser, weights?: number[]) {
  let weight = weights?.length && exercise.isWeightsRequired ? weights[0] : 0;

  if (user.defaultWeights && exercise._id && Object.hasOwn(user.defaultWeights, exercise._id)) {
    weight = user.defaultWeights[exercise._id];
  }

  return weight;
}

export function addSetToExercises(exercises: IExerciseChoosen[]) {
  const set = exercises.slice(-2).map((exercise) => {
    return { ...exercise, _id: createTempId() };
  });

  return [...exercises, ...set];
}

export function updateExercisesIndex(exercises: IExerciseChoosen[], index: number) {
  if (index <= 0 || index >= exercises.length) return exercises;

  const updatedExercises = [...exercises];
  const exerciseToMove = updatedExercises[index];

  updatedExercises.splice(index, 1);
  updatedExercises.splice(index - 1, 0, exerciseToMove);

  return updatedExercises;
}

export function updateExerciseField<T extends keyof IExerciseDone>(
  exercises: IExerciseChoosen[],
  field: T,
  value: IExerciseDone[T],
  id?: string
): IExerciseChoosen[] {
  return exercises.map((exercise) => (exercise._id === id ? { ...exercise, [field]: value } : exercise));
}

export function generateExerciseChoosen(repeats: number, weight?: number, exercise?: IExercise): IExerciseChoosen {
  return {
    _id: createTempId(),
    exercise: {
      _id: exercise?._id || '',
      title: exercise?.title || '',
      title_en: exercise?.title_en || '',
      muscles: exercise?.muscles || [],
      isWeights: exercise?.isWeights || false,
      isWeightsRequired: exercise?.isWeightsRequired || false,
      equipmentForWeight: exercise?.equipmentForWeight,
    },
    repeats,
    weight,
  };
}
