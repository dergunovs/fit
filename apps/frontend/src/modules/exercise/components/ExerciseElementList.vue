<template>
  <div>
    <UiFlex column>
      <TransitionGroup name="list" tag="div" :class="$style.block">
        <ExerciseElement
          v-for="(exercise, index) in props.exercises as IExerciseDone[]"
          :key="exercise._id"
          :exercise="exercise"
          :exercisesCount="props.exercises.length"
          :index="index"
          :isLast="index === props.exercises.length - 1"
          :isSetCreatable="isSetCreatable(props.exercises, index, exercise.exercise?._id)"
          :isEdit="props.isEdit"
          :isDone="exercise.isDone"
          :isPassing="props.isPassing"
          :duration="exercise.duration"
          :isFutureActivity="props.isFutureActivity"
          :isToFailure="exercise.isToFailure"
          :isHideTitle="
            isPrevExerciseSame(props.exercises, index, exercise.exercise?._id) && !props.isEdit && !props.isPassing
          "
          @delete="(id) => emit('delete', id)"
          @createSet="emit('createSet')"
          @setIndex="(updatedIndex) => emit('setIndex', updatedIndex)"
          @setRepeats="(repeats) => emit('setRepeats', repeats, exercise._id)"
          @setWeight="(weight) => emit('setWeight', weight, exercise._id)"
          data-test="exercise-element"
        >
          <slot :exercise="exercise" :index="index"></slot>
        </ExerciseElement>
      </TransitionGroup>
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { UiFlex } from 'mhz-ui';
import { IExerciseChoosen, IExerciseDone } from 'fitness-tracker-contracts';

import ExerciseElement from '@/exercise/components/ExerciseElement.vue';

import { isSetCreatable, isPrevExerciseSame } from '@/exercise/helpers';

interface IProps {
  exercises: IExerciseChoosen[] | IExerciseDone[];
  isEdit?: boolean;
  isPassing?: boolean;
  isFutureActivity?: boolean;
}

interface IEmit {
  delete: [id: string];
  createSet: [];
  setIndex: [index: number];
  setRepeats: [repeats: number, id?: string];
  setWeight: [weight: number, id?: string];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();
</script>

<style module lang="scss">
.block {
  display: flex;
  flex-direction: column;
  gap: 8px;
}
</style>

<style lang="scss">
.list-enter-active,
.list-leave-active,
.list-move {
  transition: all 300ms ease;
}

.list-enter-from,
.list-leave-to {
  opacity: 0;
  transform: translateY(-20px);
}

.list-leave-active {
  position: absolute;
}
</style>
