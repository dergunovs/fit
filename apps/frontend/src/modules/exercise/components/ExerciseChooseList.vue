<template>
  <UiFlex column gap="16" :class="$style.list">
    <UiFlex gap="4">
      <button
        v-for="muscle in muscleFilters"
        :key="muscle._id"
        @click="setCurrentMuscle(muscle._id)"
        type="button"
        :class="$style.button"
        :data-current="muscle._id === currentMuscle"
        data-test="muscle"
      >
        {{ muscle[localeField('title', locale)] }}
      </button>
    </UiFlex>

    <div>
      <UiField :label="t('exercise.filterByTitle')">
        <UiInput v-model="title" data-test="exercise-title" />
      </UiField>
    </div>

    <UiFlex column>
      <UiSpoiler
        v-model="exerciseSpoilers[index]"
        :title="exercise[localeField('title', locale)]"
        v-for="(exercise, index) in filterExercisesByTitleAndMuscle(props.exercises, title, currentMuscle, user)"
        :key="exercise._id"
        data-test="exercise-spoiler"
      >
        <ExerciseChooseElement
          v-if="user"
          :exercise="exercise"
          :user="user"
          @add="(choosenExercises) => emit('choose', choosenExercises)"
          data-test="exercise-choose-element"
        />
      </UiSpoiler>
    </UiFlex>
  </UiFlex>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import { IExercise, IExerciseChoosen } from 'fitness-tracker-contracts';
import { UiField, UiFlex, UiInput, UiSpoiler } from 'mhz-ui';
import { localeField } from 'mhz-helpers';

import ExerciseChooseElement from '@/exercise/components/ExerciseChooseElement.vue';

import { useAuthCheck } from '@/auth/composables';
import { muscleService } from '@/muscle/services';
import { filterExercisesByTitleAndMuscle } from '@/exercise/helpers';
import { useTI18n } from '@/common/composables';

interface IProps {
  exercises: IExercise[];
}

interface IEmit {
  choose: [choosenExercises: IExerciseChoosen[]];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t, locale } = useTI18n();
const { user } = useAuthCheck();

const exerciseSpoilers = ref([]);

const currentMuscle = ref<string>('');
const title = ref('');

const { data: muscles } = muscleService.getAll();

const muscleFilters = computed(() => [{ _id: '', title: t('all'), title_en: t('all') }, ...(muscles.value || [])]);

function setCurrentMuscle(id?: string) {
  currentMuscle.value = id || '';
  exerciseSpoilers.value = [];
}
</script>

<style module lang="scss">
.list {
  min-height: 53dvh;
}

.button {
  padding: 4px 1px;
  font-size: 0.875rem;
  font-weight: 700;
  color: var(--color-gray-dark);
  text-decoration: underline;
  cursor: pointer;
  background: none;
  border: none;

  &[data-current='true'] {
    color: var(--color-primary);
  }
}
</style>
