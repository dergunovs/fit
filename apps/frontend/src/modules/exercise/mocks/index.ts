import { vi } from 'vitest';
import {
  TDeleteExerciseDTO,
  TGetExercisesDTO,
  TPostExerciseDataDTO,
  TPostExerciseDTO,
  TUpdateExerciseDataDTO,
  TUpdateExerciseDTO,
} from 'fitness-tracker-contracts';

import { exerciseService } from '@/exercise/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { EXERCISE_FIXTURE, EXERCISE_FIXTURE_CUSTOM, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetExercise = vi.spyOn(exerciseService, 'getOne').mockReturnValue(mockQueryReply(EXERCISE_FIXTURE));

const mockGetExercisesData: TGetExercisesDTO = { data: EXERCISES_FIXTURE, total: EXERCISES_FIXTURE.length };

const spyGetExercises = vi
  .spyOn(exerciseService, 'getMany')
  .mockImplementation(() => mockQueryReply(mockGetExercisesData));

const spyGetExercisesAll = vi
  .spyOn(exerciseService, 'getAll')
  .mockImplementation(() => mockQueryReply(EXERCISES_FIXTURE));

const spyGetExercisesCustom = vi
  .spyOn(exerciseService, 'getCustom')
  .mockImplementation(() => mockQueryReply([EXERCISE_FIXTURE_CUSTOM]));

const spyCreateExercise = vi.fn();
const spyUpdateExercise = vi.fn();
const spyDeleteExercise = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(exerciseService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostExerciseDTO, TPostExerciseDataDTO>(spyCreateExercise);
});

vi.spyOn(exerciseService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateExerciseDTO, TUpdateExerciseDataDTO>(spyUpdateExercise);
});

vi.spyOn(exerciseService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteExerciseDTO, string>(spyDeleteExercise);
});

export {
  spyGetExercise,
  spyGetExercises,
  spyGetExercisesCustom,
  spyGetExercisesAll,
  spyCreateExercise,
  spyUpdateExercise,
  spyDeleteExercise,
  mockOnSuccess,
  mockGetExercisesData,
};
