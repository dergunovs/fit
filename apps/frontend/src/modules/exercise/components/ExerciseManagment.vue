<template>
  <div>
    <UiModal
      :modelValue="props.isShowModal"
      @update:modelValue="(value) => updateModal(value)"
      isScrollable
      data-test="exercise-modal"
    >
      <ExerciseChooseList
        v-if="exercisesAll?.length"
        :exercises="exercisesAll"
        @choose="addExercises"
        data-test="exercise-choose-list"
      />
    </UiModal>

    <UiFlex v-if="props.modelValue?.length" column>
      <h3>{{ t('exercise.many') }}</h3>

      <ExerciseElementList
        :exercises="props.modelValue"
        isEdit
        @delete="deleteExercise"
        @createSet="createSet"
        @setIndex="updateIndex"
        @setRepeats="updateRepeats"
        @setWeight="updateWeight"
        data-test="exercise-element-list"
      />
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { UiFlex, UiModal } from 'mhz-ui';
import { IExerciseChoosen } from 'fitness-tracker-contracts';

import ExerciseChooseList from '@/exercise/components/ExerciseChooseList.vue';
import ExerciseElementList from '@/exercise/components/ExerciseElementList.vue';

import { useTI18n } from '@/common/composables';
import { exerciseService } from '@/exercise/services';
import {
  addSetToExercises,
  updateExercisesIndex,
  updateExercisesRepeats,
  updateExercisesWeight,
} from '@/exercise/helpers';

interface IProps {
  isShowModal: boolean;
  modelValue?: IExerciseChoosen[];
}

interface IEmit {
  updateModal: [value: boolean];
  'update:modelValue': [exercises: IExerciseChoosen[]];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t } = useTI18n();

const { data: exercisesAll } = exerciseService.getAll();

function addExercises(exercises: IExerciseChoosen[]) {
  const updatedExercises = props.modelValue?.length ? [...props.modelValue, ...exercises] : exercises;

  emit('update:modelValue', updatedExercises);

  updateModal(false);
}

function deleteExercise(idToDelete: string) {
  const updatedExercises = props.modelValue?.filter((exercise) => exercise._id !== idToDelete) || [];

  emit('update:modelValue', updatedExercises);
}

function createSet() {
  if (!props.modelValue) return;

  const updatedExercises = addSetToExercises(props.modelValue);

  emit('update:modelValue', updatedExercises);
}

function updateIndex(updatedIndex: number) {
  if (!props.modelValue) return;

  const updatedExercises = updateExercisesIndex(props.modelValue, updatedIndex);

  emit('update:modelValue', updatedExercises);
}

function updateRepeats(repeats: number, id?: string) {
  if (!props.modelValue || !id) return;

  const updatedExercises = updateExercisesRepeats(props.modelValue, repeats, id);

  emit('update:modelValue', updatedExercises);
}

function updateWeight(weight: number, id?: string) {
  if (!props.modelValue || !id) return;

  const updatedExercises = updateExercisesWeight(props.modelValue, weight, id);

  emit('update:modelValue', updatedExercises);
}

function updateModal(value: boolean) {
  emit('updateModal', value);
}
</script>
