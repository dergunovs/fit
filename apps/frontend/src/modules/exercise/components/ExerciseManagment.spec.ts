import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { UiModal } from 'mhz-ui';
import { dataTest } from 'mhz-helpers';

import ExerciseManagment from './ExerciseManagment.vue';
import ExerciseChooseList from '@/exercise/components/ExerciseChooseList.vue';
import ExerciseElementList from '@/exercise/components/ExerciseElementList.vue';

import { wrapperFactory } from '@/common/test';
import { EXERCISES_CHOOSEN_FIXTURE, EXERCISE_CHOOSEN_FIXTURE, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { spyGetExercisesAll } from '@/exercise/mocks';
import { addSetToExercises, updateExercisesIndex, updateExerciseField } from '@/exercise/helpers';

const ID = EXERCISES_CHOOSEN_FIXTURE[0]._id;

const modal = dataTest('exercise-modal');
const chooseList = dataTest('exercise-choose-list');
const exercises = dataTest('exercise-element-list');

let wrapper: VueWrapper<InstanceType<typeof ExerciseManagment>>;

beforeEach(() => {
  wrapper = wrapperFactory(ExerciseManagment, { isShowModal: false, modelValue: EXERCISES_CHOOSEN_FIXTURE });
});

enableAutoUnmount(afterEach);

describe('ExerciseManagment', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ExerciseManagment)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('gets exercises data', async () => {
    expect(spyGetExercisesAll).toBeCalledTimes(1);
    expect(wrapper.findComponent<typeof ExerciseChooseList>(chooseList).props('exercises')).toStrictEqual(
      EXERCISES_FIXTURE
    );
  });

  it('hides modal by default', async () => {
    expect(wrapper.find(modal).attributes('modelvalue')).toStrictEqual('false');
  });

  it('emits update modal', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('updateModal');

    wrapper.findComponent<typeof UiModal>(modal).vm.$emit('update:modelValue', true);

    expect(wrapper.emitted('updateModal')).toHaveLength(1);
    expect(wrapper.emitted()['updateModal'][0]).toStrictEqual([true]);
  });

  it('passes exercises to choosen list', async () => {
    expect(wrapper.findComponent<typeof ExerciseElementList>(exercises).props('exercises')).toStrictEqual(
      EXERCISES_CHOOSEN_FIXTURE
    );
  });

  it('adds exercises', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('update:modelValue');
    expect(wrapper.emitted()).not.toHaveProperty('updateModal');

    wrapper.findComponent<typeof ExerciseChooseList>(chooseList).vm.$emit('choose', [EXERCISE_CHOOSEN_FIXTURE]);

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual([
      [...EXERCISES_CHOOSEN_FIXTURE, EXERCISE_CHOOSEN_FIXTURE],
    ]);

    expect(wrapper.emitted('updateModal')).toHaveLength(1);
    expect(wrapper.emitted()['updateModal'][0]).toStrictEqual([false]);
  });

  it('deletes exercise', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('update:modelValue');

    wrapper.findComponent<typeof ExerciseElementList>(exercises).vm.$emit('delete', ID);

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual([[EXERCISES_CHOOSEN_FIXTURE[1]]]);
  });

  it('adds set', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('update:modelValue');

    wrapper.findComponent<typeof ExerciseElementList>(exercises).vm.$emit('createSet');

    const updatedExercises = addSetToExercises(EXERCISES_CHOOSEN_FIXTURE);

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual([updatedExercises]);
  });

  it('sets new exercise index', async () => {
    const INDEX = 2;

    expect(wrapper.emitted()).not.toHaveProperty('update:modelValue');

    wrapper.findComponent<typeof ExerciseElementList>(exercises).vm.$emit('setIndex', INDEX);

    const updatedExercises = updateExercisesIndex(EXERCISES_CHOOSEN_FIXTURE, INDEX);

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual([updatedExercises]);
  });

  it('updates repeats', async () => {
    const REPEATS = 11;

    expect(wrapper.emitted()).not.toHaveProperty('update:modelValue');

    wrapper.findComponent<typeof ExerciseElementList>(exercises).vm.$emit('setRepeats', REPEATS, ID);

    const updatedExercises = updateExerciseField(EXERCISES_CHOOSEN_FIXTURE, 'repeats', REPEATS, ID as string);

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual([updatedExercises]);
  });

  it('updates weight', async () => {
    const WEIGHT = 44;

    expect(wrapper.emitted()).not.toHaveProperty('update:modelValue');

    wrapper.findComponent<typeof ExerciseElementList>(exercises).vm.$emit('setWeight', WEIGHT, ID);

    const updatedExercises = updateExerciseField(EXERCISES_CHOOSEN_FIXTURE, 'weight', WEIGHT, ID as string);

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted()['update:modelValue'][0]).toStrictEqual([updatedExercises]);
  });
});
