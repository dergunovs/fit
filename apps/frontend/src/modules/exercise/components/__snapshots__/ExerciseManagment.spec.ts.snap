// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExerciseManagment > matches snapshot 1`] = `
"<div>
  <div modelvalue="false" isscrollable="" data-test="exercise-modal">
    <exercise-choose-list-stub exercises="[object Object],[object Object],[object Object]" data-test="exercise-choose-list"></exercise-choose-list-stub>
  </div>
  <div column="">
    <h3>Упражнения</h3>
    <exercise-element-list-stub exercises="[object Object],[object Object]" isedit="true" ispassing="false" isfutureactivity="false" data-test="exercise-element-list"></exercise-element-list-stub>
  </div>
</div>"
`;
